src/Microsoft.AspNet.Mvc.ModelBinding/Validation/DefaultBodyModelValidator.cs

```diff
@@ -138,66 +130,69 @@ private bool ValidateElements(IEnumerable model, ValidationContext validationCon
                 if (element != null || anyValidatorsDefined)
                 {
                     elementMetadata.Model = element;
-
-                    if (!ValidateNonVisitedNodeAndChildren(elementMetadata, validationContext, validators))
+                    var elementKey = ModelBindingHelper.CreateIndexModelName(currentKey, index);
+                    if (!ValidateNonVisitedNodeAndChildren(elementKey, elementMetadata, validationContext, validators))
                     {
                         isValid = false;
                     }
                 }
 
-                elementScope.Index++;
+                index++;
             }
 
-            validationContext.KeyBuilders.Pop();
             return isValid;
         }
 
         // Validates a single node (not including children)
         // Returns true if validation passes successfully
         private static bool ShallowValidate(
+            string modelKey,
             ModelMetadata metadata,
             ValidationContext validationContext,
             [NotNull] IEnumerable<IModelValidator> validators)
         {
             var isValid = true;
-            string modelKey = null;
 
             // When the are no validators we bail quickly. This saves a GetEnumerator allocation.
             // In a large array (tens of thousands or more) scenario it's very significant.
```

Comment:
You can probably safely kill old 'optimizations' like this. This was written within the constraints of some api signature that we were probably locked to.

You could for instance just turn `validators` into an `IReadOnlyList<>` and simplify all of this business. We should remove the cruft and then profile to find hotspots. It's likely we can do better since we have the freedom to change apis.